### PR TITLE
feat: igraph_is_simple() uses the cache in a more granular way

### PR DESCRIPTION
**Pull request by szhorvat**
Asking for a review, since this touches code that sets the cache.

This PR makes more granular use of the cache in `igraph_is_simple()`. Previously, it would only use the cache when the status of _both_ HAS_LOOP and HAS_MULTI was known.

However, if either of these is true, we already know that the graph is not simple without having to check the other.

Similarly, we can set set the cache in a more granular way when either was found.

The current plan for property deduction is to NOT include a new flag for IS_SIMPLE, therefore we can handle this directly in the `igraph_is_simple()` function instead.

### CODE REVIEW

File: src/properties/multiplicity.c [hunk 1]
```diff
@@ -50,46 +50,74 @@ igraph_error_t igraph_is_simple(const igraph_t *graph, igraph_bool_t *res) {
...
+        }
         igraph_vector_int_destroy(&neis);
         IGRAPH_FINALLY_CLEAN(1);
     }
 
-    /* If the graph turned out to be simple, we can cache that it has no loop
-     * and no multiple edges */
-    if (*res) {
+    if (known_loop) {
         igraph_i_property_cache_set_bool(graph, IGRAPH_PROP_HAS_LOOP, false);
```
Comment (ntamas) [inline src/properties/multiplicity.c:117]: Should `false` be `has_loop` here instead?

File: src/properties/multiplicity.c [hunk 2]
```diff
@@ -50,46 +50,74 @@ igraph_error_t igraph_is_simple(const igraph_t *graph, igraph_bool_t *res) {
...
+    }
+
+    if (known_multi) {
         igraph_i_property_cache_set_bool(graph, IGRAPH_PROP_HAS_MULTI, false);
```
Comment (ntamas) [inline src/properties/multiplicity.c:121]: Should `false` be `known_multi` here instead?

File: src/properties/multiplicity.c (hunk 1 above)
Comment (szhorvat) [reply src/properties/multiplicity.c]: Yes, this is why I asked for a review.

File: src/properties/multiplicity.c [hunk 3]
```diff
@@ -50,47 +50,75 @@ igraph_error_t igraph_is_simple(const igraph_t *graph, igraph_bool_t *res) {
...
         IGRAPH_FINALLY_CLEAN(1);
     }
 
-    /* If the graph turned out to be simple, we can cache that it has no loop
-     * and no multiple edges */
-    if (*res) {
-        igraph_i_property_cache_set_bool(graph, IGRAPH_PROP_HAS_LOOP, false);
-        igraph_i_property_cache_set_bool(graph, IGRAPH_PROP_HAS_MULTI, false);
+    if (known_loop) {
+        igraph_i_property_cache_set_bool(graph, IGRAPH_PROP_HAS_LOOP, has_loop);
```
Comment (ntamas) [inline src/properties/multiplicity.c:117]: This branch will never be hit if `vc == 0 || ec == 0` and neither the other. Is this intentional? If it is, then I think the current commit is fine.

**Comment by szhorvat** [comment]
@ntamas This is good to go now.

**Comment by ntamas** [comment]
LGTM. One last CI check with an updated `master` and if it passes, I'll merge. Thanks!

**Pull request merged**